src/ui/reducers/hitCounts.ts

```diff
@@ -0,0 +1,173 @@
+import { createEntityAdapter, createSlice, EntityState, PayloadAction } from "@reduxjs/toolkit";
+import { Location } from "@replayio/protocol";
+import { UIThunkAction } from "ui/actions";
+import { UIState } from "ui/state";
+import { rangeForFocusRegion } from "ui/utils/timeline";
+import { getCorrespondingSourceIds, getSourceDetails } from "./sources";
+import { getFocusRegion } from "./timeline";
+import { fetchProtocolHitCounts } from "protocol/thread/hitCounts";
+import sortBy from "lodash/sortBy";
+import { getSelectedSourceId } from "devtools/client/debugger/src/selectors";
+
+export interface HitCount {
+  location: Location;
+  hits: number;
+}
+
+export interface SourceHitCounts {
+  error?: string;
+  hitCounts?: HitCount[];
+  id: string;
+  status: LoadingState;
+}
+
+export type HitCountsState = EntityState<SourceHitCounts>;
+
+export enum LoadingState {
+  LOADING = "loading",
+  LOADED = "loaded",
+  ERRORED = "errored",
+}
+
+const hitCountsAdapter = createEntityAdapter<SourceHitCounts>();
+const initialState = hitCountsAdapter.getInitialState();
+
+const hitCountsSlice = createSlice({
+  name: "hitCounts",
+  initialState,
+  reducers: {
+    hitCountsRequested: (state, action: PayloadAction<string>) => {
+      hitCountsAdapter.upsertOne(state, {
+        id: action.payload,
+        status: LoadingState.LOADING,
+      });
+    },
+    hitCountsReceived: (state, action: PayloadAction<{ id: string; hitCounts: HitCount[] }>) => {
+      hitCountsAdapter.upsertOne(state, {
+        ...action.payload,
+        status: LoadingState.LOADED,
+      });
+    },
+    hitCountsFailed: (state, action: PayloadAction<{ id: string; error: string }>) => {
+      hitCountsAdapter.upsertOne(state, {
+        id: action.payload.id,
+        status: LoadingState.ERRORED,
+        error: action.payload.error,
+      });
+    },
+  },
+});
+
+const hitCountsSelectors = hitCountsAdapter.getSelectors<UIState>(
```

Comment:
:speech_balloon: don't think we need the initial generic, just the `state: UIState`, but not an issue :)